ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/execution/PullQueryExecutor.java

```diff
@@ -224,36 +234,73 @@ public PullQueryResult execute(
           .getMaterialization(queryId, contextStacker)
           .orElseThrow(() -> notMaterializedException(getSourceName(analysis)));
 
-      final Struct key = asKeyStruct(whereInfo.keyBound, query.getPhysicalSchema());
+      final List<Optional<KsqlNode>> sourceNodes = new ArrayList<>();
+      final List<List<?>> tableRows = new ArrayList<>();
+      final List<LogicalSchema> schemas = new ArrayList<>();
+      final List<Future<PullQueryResult>> futures = new ArrayList<>();
+      final List<Struct> keys = whereInfo.keysBound.stream()
+          .map(keyBound -> asKeyStruct(keyBound, query.getPhysicalSchema()))
+          .collect(ImmutableList.toImmutableList());
+      final List<KsqlNodeList> nodeLists = mat.locator().locate(
+          keys,
+          routingOptions,
+          routingFilterFactory
+      );
 
-      final PullQueryContext pullQueryContext = new PullQueryContext(
-          key,
-          mat,
-          analysis,
-          whereInfo,
-          queryId,
-          contextStacker,
-          pullQueryMetrics);
+      for (KsqlNodeList ksqlNodeList : nodeLists) {
+        final PullQueryContext pullQueryContext = new PullQueryContext(
+            ksqlNodeList.getKeys(),
+            mat,
+            analysis,
+            whereInfo,
+            queryId,
+            contextStacker,
+            pullQueryMetrics);
+
+        futures.add(executorService.submit(() -> handlePullQuery(
+            statement,
+            executionContext,
+            serviceContext,
+            pullQueryContext,
+            ksqlNodeList.getNodes(),
+            routingOptions
+        )));
 
-      final PullQueryResult result = handlePullQuery(
-          statement,
-          executionContext,
-          serviceContext,
-          pullQueryContext,
-          routingOptions
-      );
+      }
+      for (Future<PullQueryResult> future : futures) {
```

Comment:
is there any requirement that the pull query result is reproducible? it seems like we add rows to the result based on the order of the nodes that we request, but those nodes can host different partitions at different times. I think it's reasonable to not guarantee ordering, but food for thought (and probably should be documented)

Reply:
Yes, you're right that there are no guarantees about ordering.  I'll document that in the code and consider documenting elsewhere as well.